text/0000-crate-attr.md

```diff
@@ -0,0 +1,113 @@
+
+The argument to `--crate-attr` is treated as-if it were surrounded by `#![ ]`, i.e. it must be an inner attribute and it cannot include multiple attributes, nor can it be any grammar production other than an [`Attr`].
+
+If the attribute is already present in the source code, it behaves exactly as it would if duplicated twice in the source.
+For example, duplicating `no_std` is idempotent; duplicating `crate_type` generates both types; and duplicating `crate_name` is idempotent if the names are the same and a hard error otherwise.
+It is suggested, but not required, that the implementation not warn on idempotent attributes, even if it would normally warn that duplicate attributes are unused.
+
+[shebang parsing]: https://doc.rust-lang.org/nightly/reference/input-format.html#shebang-removal
+[`Attr`]: https://doc.rust-lang.org/nightly/reference/attributes.html
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It makes it harder for Rust developers to know whether it's idiomatic to use flags or attributes.
+In practice, this has not be a large drawback for `crate_name` and `crate_type`, although for lints perhaps a little more so since they were only recently stabilized in Cargo.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- We could require `--crate-attr=#![foo]` instead. This is more verbose and requires extensive shell quoting, for not much benefit.
+- We could disallow comments in the attribute. This perhaps makes the design less surprising, but complicates the implementation for little benefit.
+- We could add a syntax for passing multiple attributes in a single CLI flag. We would have to find a syntax that avoids ambiguity *and* that does not mis-parse the data inside string literals (i.e. picking a fixed string, such as `|`, would not work because it has to take quote nesting into account). This greatly complicates the implementation for little benefit.
+
+This cannot be done in a library or macro. It can be done in an external tool, but only by modifying the source in place, which requires first parsing it, and in general is much more brittle than this approach (for example, preventing the argument from injecting a unterminated block comment, or from injecting a non-attribute grammar production, becomes much harder).
+
+In the author's opinion, having source injected via this mechanism does not make code any harder to read than the existing flags that are already stable (in particular `-C panic` and `--edition` come to mind).
+
+# Prior art
+[prior-art]: #prior-art
+
+- HTML allows `<meta http-equiv=...>` to emulate headers, which is very useful for using hosted infra where one does not control the server.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should this interact with doctests? Does it apply to the crate being tested or to the generated test?
```

Comment:
after thinking about this, i think this should always apply to the crate being tested. to apply to the generated test people can use `--crate-attr=doc(test(attr(...)))` (which is a bit of a mouthful, but, at least it's unambiguous.)

Reply:
such behavior would deviate from the semantics of existing flags like `--warn` which apply to *both* the crate being tested and its doctests.

Reply:
hmm, ok. i will change the RFC to specify it should apply to both.

cc @rust-lang/rustdoc

Reply:
i actually expanded this into something more complicated; see the RFC for details.

Reply:
From what I can tell `rustdoc --test --deny=...` does not apply to either the crate being tested or its doctests. `rustdoc --test -Zcrate-attr` seems to only apply to the generated doctests crates, so `-Zcrate-attr=doc(test(attr(...)))` does not currently work.

Reply:
both of those seem like implementation bugs to me. i am happy to open an issue on rl/r if this RFC is merged.